build-scripts/build_tests.py

```diff
@@ -0,0 +1,237 @@
+                if scenario.matches_platform(benchmark_cpes):
+                    rule_output_path.mkdir(parents=True, exist_ok=True)
+                    test_output_path = rule_output_path / test.name
+                    _write_path(file_contents, test_output_path)
+                    logger.debug('Wrote scenario %s for rule %s', test.name, rule_id)
+                else:
+                    logger.debug('Skipping scenario %s for rule %s has it not applicable',
+                                 test.name, rule_id)
+
+
+def _get_benchmark_cpes(env_yaml):
+    benchmark_cpes = set()
+    for cpe in env_yaml["cpes"]:
+        for _, data in cpe.items():
+            benchmark_cpes.add(data["name"])
+    return benchmark_cpes
+
+
+def _get_rules_in_profile(built_profiles_root):
+    rules_in_profile = list()
+    for profile_file in built_profiles_root.iterdir():  # type: pathlib.Path
+        if not profile_file.name.endswith('.profile'):
+            continue
+        profile_data = ssg.yaml.open_raw(str(profile_file.absolute()))
+        for selection in profile_data["selections"]:
+            if '=' not in selection:
+                rules_in_profile.append(selection)
+    return rules_in_profile
+
+
+def main() -> int:
+    args = _create_arg_parser().parse_args()
+    logging.basicConfig(level=logging.getLevelName(args.log_level))
+    env_yaml = ssg.environment.open_environment(
+        args.build_config_yaml, args.product_yaml)
+
+    root_path = pathlib.Path(args.root).resolve().absolute()
+    output_path = pathlib.Path(args.output).resolve().absolute()
+    resolved_rules_dir = pathlib.Path(args.resolved_rules_dir)
+    if not resolved_rules_dir.exists() or not resolved_rules_dir.is_dir():
+        logging.error("Unable to find product at %s", str(resolved_rules_dir))
+        logging.error("Is the product built?")
+        return 1
+
+    output_path.mkdir(parents=True, exist_ok=True)
+    _copy_and_process_shared(env_yaml, output_path, root_path)
+    benchmark_cpes = _get_benchmark_cpes(env_yaml)
+
+    built_profiles_root = resolved_rules_dir.parent / "profiles"
+    rules_in_profiles = _get_rules_in_profile(built_profiles_root)
+
+    templates_root = root_path / "shared" / "templates"
+    all_resolved_rules = list(resolved_rules_dir.iterdir())
+    processes = list()
+    for chunk in range(args.jobs):
+        process_args = (benchmark_cpes, env_yaml, output_path,
+                        all_resolved_rules[chunk::args.jobs], templates_root, rules_in_profiles,)
+        process = multiprocessing.Process(target=_process_rules, args=process_args)
+        processes.append(process)
+        process.start()
+    for process in processes:
+        process.join()
+    # Write a file for CMake
+    # So we don't have dependency on a folder
+    done_file: pathlib.Path = output_path / ".test_done"
+    done_file.touch()
+    return 0
+
+
+if __name__ == "__main__":
+    raise SystemExit(main())
```

Comment:
I would honestly just let python interpreter defaults handle exit code - raise an exception if something bad happens, otherwise let it finish cleanly.

If you really want an exception-less exit on `resolved_rules_dir` not existing (seems like an arbitrary choice?), just do `sys.exit(1)` in that one case. `sys` will take care of the exit-related exception handling.

Reply:
I can remove to `sys.exit`. However, under the covers this is how `sys.exit` works.

Reply:
I guess either way works. I remember some problems with raising `SystemExit()` with non-CPython intepreters, but we don't use those here, and those problems are probably fixed by now anyway. I would still leave the `0` exit to be done by default instead of `SystemExit(0)`, but that's probably bikeshedding.